### PR TITLE
[mesheryctl] Ensuring that mesheryctl system check works perfectly

### LINKED ISSUE #8425: [mesheryctl] Ensuring that mesheryctl system check works perfectly
_See [mesheryctl Command Tracker](https://bit.ly/3dqXy1q) for current status of commands._

#### Desired Behavior

Ensure that system check works in all situations and that these functions are pervasively reused in other commands.

#### Mockups

---

#### Contributor Resources

- ⌨️ [mesheryctl Contributing Guide](https://github.com/meshery/meshery/blob/master/mesheryctl/README.md)
- [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)
- 🙋🏾🙋🏼 Questions: [Layer5 Discussion Forum](https://discuss.layer5.io) and [Layer5 Community Slack](http://slack.layer5.io)

### LINKED ISSUE #8590: [mesheryctl] system start command dysfunctional for latest kubectl version
_See [mesheryctl Command Tracker](https://bit.ly/3dqXy1q) for current status of commands._

### Current Behavior

For latest kubernetes version `mesheryctl system start -p kubernetes` is throwing error
image

### Desired Behavior

mesheryctl should work with latest kubectl version

### Environment

- Host OS: Mac Linux Windows
- Meshery Server Version: stable-v
- Meshery Client Version: stable-v
- Platform: Kubernetes

---

### Contributor [Guides](https://docs.meshery.io/project/contributing) and [Handbook](https://layer5.io/community/handbook)
- 📚 [mesheryctl Contributing Guide](https://github.com/meshery/meshery/blob/master/mesheryctl/README.md)
- 📺 [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- ⌨️ [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- ⌨️ [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)
- 🙋🏾🙋🏼 Questions: [Discussion Forum](https://discuss.layer5.io) and [Community Slack](http://slack.layer5.io)

### PR DESCRIPTION
**Pull request by suhail34**
**Notes for Reviewers**

This PR fixes #8425
This PR ensures that before running any of Mesheryctl command (like app, filter, etc) a healthcheck is being done for the readiness of meshery some improvements also needed to be done with respect to healthcheck of adapters as by default we don't run any of adapters also fixes #8590 

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by hexxdump** [comment]
> **Notes for Reviewers**
> 
> This PR fixes #8425
> 
> **[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
> 
>     * [ ]  Yes, I signed my commits.

Can you please add few more details on defining the problem in detail and how the solution handles the issues?
Problem report also does not list much.

File: mesheryctl/pkg/utils/error.go [hunk 1]
```diff
@@ -309,3 +310,8 @@ func ErrInvalidNameOrID(err error) error {
 func ErrInvalidAPIResponse(err error) error {
 	return errors.New(ErrInvalidAPIResponseCode, errors.Fatal, []string{"Invalid API response encountered"}, []string{"Invalid API response encountered", err.Error()}, []string{}, []string{})
 }
+
+func ErrGetKubectlClientVersion(err error) error {
+	return errors.New(ErrGetKubectlClientVersionCode, errors.Alert, []string{"Error fetching kubectl client version"}, []string{"Error fetching kubectl client version", err.Error()},
+		[]string{"You might not have the kubectl binary to fetch the version"}, []string{"Install the minimum kubectl binary from the kubernetes docs"})
```
Comment (hexxdump) [inline mesheryctl/pkg/utils/error.go:316]:
```suggestion
		[]string{"You do not have the kubectl binary or you do not have permission to run kubectl command to fetch the version"}, []string{"Install the kubectl binary by referring to instructions in the kubernetes docs"})
```

File: mesheryctl/internal/cli/root/app/app.go [hunk 2]
```diff
@@ -42,10 +43,27 @@ var AppCmd = &cobra.Command{
 // Base command
 mesheryctl app [subcommand]
 	`,
-	RunE: func(cmd *cobra.Command, args []string) error {
-		if len(args) == 0 {
-			return cmd.Help()
+	PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
```
Comment (hexxdump) [inline mesheryctl/internal/cli/root/app/app.go:46]: This code is repeated multiple times, will it be possible to move it to utils package and all other mesheryctl sub commands make use of this function ?

Comment (suhail34) [reply mesheryctl/internal/cli/root/app/app.go:45]: @hexxdump  done

**Comment by leecalcote** [comment]
Some good review happening here, @hexxdump 👏

File: .github/workflows/error-codes-updater.yaml [hunk 3]
```diff
@@ -44,7 +44,7 @@ jobs:
         run: |
           cd ./mesheryctl
           ~/errorutil update -d . -i ./helpers -o ./helpers
-          $output=$(~/errorutil -d . analyze -i ./helpers -o ./helpers | wc -c)
+          output=$(~/errorutil -d . analyze -i ./helpers -o ./helpers | wc -c)
```
Comment (hexxdump) [inline .github/workflows/error-codes-updater.yaml:47]: good catch!

File: mesheryctl/pkg/utils/error.go [hunk 4]
```diff
@@ -11,21 +11,22 @@ import (
...
+	ErrCreatingRequestCode         = "1006"
+	ErrMarhallingCode              = "1007"
+	ErrReadResponseBodyCode        = "1008"
+	ErrParsingUrlCode              = "1009"
+	ErrNotFoundCode                = "1010"
+	ErrUnauthenticatedCode         = "1011"
+	ErrInvalidFileCode             = "1012"
+	ErrInvalidNameOrIDCode         = "1013"
+	ErrInvalidAPIResponseCode      = "1014"
+	ErrGetKubectlClientVersionCode = "replace_me"
```
Comment (hexxdump) [inline mesheryctl/pkg/utils/error.go:29]: Replace the error code with proper one.

**Comment by leecalcote** [comment]
@suhail34 do you mind looking into these merge conflicts?

**Comment by AugustasV** [comment]
@suhail34 https://github.com/meshery/meshery/actions/runs/6225823323/job/16897055329?pr=8602 what is that error mean?

```time="2023-09-18T17:17:59Z" level=error msg="duplicate error code name 'ErrProviderInfoCode'"
time="2023-09-18T17:17:59Z" level=error msg="duplicate error code name 'ErrInvalidNameOrIDCode'"
time="2023-09-18T17:17:59Z" level=error msg="duplicate error details for error name 'ErrInvalidNameOrIDCode' and code '[11](https://github.com/meshery/meshery/actions/runs/6225823323/job/16897055329?pr=8602#step:6:12)76'"
time="2023-09-18T17:17:59Z" level=warning msg="no error details found for error name 'ErrAppFlagCode' and code '1085'"
time="2023-09-18T17:17:59Z" level=warning msg="no error details found for error name 'ErrInvalidAPICallCode' and code '1185'"
time="2023-09-18T17:17:59Z" level=error msg="duplicate error details for error name 'ErrProviderInfoCode' and code '1[15](https://github.com/meshery/meshery/actions/runs/6225823323/job/16897055329?pr=8602#step:6:16)9'"
time="2023-09-18T17:17:59Z" level=error msg="duplicate error details for error name 'ErrInvalidNameOrIDCode' and code '1084'"
time="2023-09-18T17:17:59Z" level=warning msg="no error details found for error name 'ErrValidProviderCode' and code '1[16](https://github.com/meshery/meshery/actions/runs/6225823323/job/16897055329?pr=8602#step:6:17)0'"
Error: Process completed with exit code 1.

**Comment by suhail34** [comment]
> @suhail34 https://github.com/meshery/meshery/actions/runs/6225823323/job/16897055329?pr=8602 what is that error mean?
> 
> ```
> time="2023-09-18T17:17:59Z" level=error msg="duplicate error code name 'ErrInvalidNameOrIDCode'"
> time="2023-09-18T17:17:59Z" level=error msg="duplicate error details for error name 'ErrInvalidNameOrIDCode' and code '[11](https://github.com/meshery/meshery/actions/runs/6225823323/job/16897055329?pr=8602#step:6:12)76'"
> time="2023-09-18T17:17:59Z" level=warning msg="no error details found for error name 'ErrAppFlagCode' and code '1085'"
> time="2023-09-18T17:17:59Z" level=warning msg="no error details found for error name 'ErrInvalidAPICallCode' and code '1185'"
> time="2023-09-18T17:17:59Z" level=error msg="duplicate error details for error name 'ErrProviderInfoCode' and code '1[15](https://github.com/meshery/meshery/actions/runs/6225823323/job/16897055329?pr=8602#step:6:16)9'"
> time="2023-09-18T17:17:59Z" level=error msg="duplicate error details for error name 'ErrInvalidNameOrIDCode' and code '1084'"
> time="2023-09-18T17:17:59Z" level=warning msg="no error details found for error name 'ErrValidProviderCode' and code '1[16](https://github.com/meshery/meshery/actions/runs/6225823323/job/16897055329?pr=8602#step:6:17)0'"
> Error: Process completed with exit code 1.
> ```

There are some duplicacy added to meshkit errors so that's why it is failing

**Comment by AugustasV** [comment]
@suhail34 I see that those errors are also in other issues. So it is good to merge? @hexxdump

**Review by hexxdump — DISMISSED** [review]
LGTM

**Comment by AugustasV** [comment]
@suhail34 it should be good to merge

**Comment by leecalcote** [comment]
@suhail34 will you address these merge conflicts?

**Comment by alphaX86** [comment]
@suhail34 there is still one change needed/to be explained in this PR....

File: mesheryctl/pkg/utils/error.go [hunk 5]
```diff
@@ -11,32 +11,33 @@ import (
...
+	ErrInvalidNameOrIDCode         = "replace_me"
+	ErrInvalidAPIResponseCode      = "replace_me"
+	ErrReadConfigFileCode          = "replace_me"
+	ErrMarshalIndentCode           = "replace_me"
+	ErrLoadConfigCode              = "replace_me"
+	ErrResponseStatusBodyCode      = "replace_me"
+	ErrResponseStatusCode          = "replace_me"
+	ErrJSONToYAMLCode              = "replace_me"
+	ErrOutFormatFlagCode           = "replace_me"
+	ErrInvalidAPICallCode          = "replace_me"
```
Comment (alphaX86) [inline mesheryctl/pkg/utils/error.go:36]: Is it good if we leave it as such?!

**Review by alphaX86 — CHANGES_REQUESTED** [review]

**Comment by AugustasV** [comment]
@suhail34 can you fix remaining comments on that PR ?

**Pull request closed without merging**